src/dfx/src/lib/operations/canister/motoko_playground.rs

```diff
@@ -0,0 +1,191 @@
+pub async fn authorize_asset_uploader(
+    env: &dyn Environment,
+    canister_id: Principal,
+    canister_timestamp: candid::Int,
+    principal_to_authorize: &Principal,
+) -> DfxResult {
+    let agent = env.get_agent().context("Failed to get HTTP agent")?;
+    let playground_canister = if let NetworkTypeDescriptor::Playground {
+        playground_canister,
+        ..
+    } = env.get_network_descriptor().r#type
+    {
+        playground_canister
+    } else {
+        bail!("Trying to authorize asset uploader on non-playground network.")
+    };
+    let canister_info = CanisterInfo {
+        id: canister_id,
+        timestamp: canister_timestamp,
+    };
+
+    let nested_arg = Encode!(&principal_to_authorize)?;
+    let call_arg = Encode!(&canister_info, &"authorize", &nested_arg)?;
+
+    let _ = agent
+        .update(&playground_canister, "callForward")
+        .with_arg(call_arg)
+        .call_and_wait()
+        .await
+        .context("Failed to call playground.")?;
+    Ok(())
+}
+
+pub async fn playground_install_code(
+    env: &dyn Environment,
+    canister_id: Principal,
+    canister_timestamp: candid::Int,
+    arg: &[u8],
+    wasm_module: &[u8],
+    mode: InstallMode,
+    is_asset_canister: bool,
+) -> DfxResult<num_bigint::BigInt> {
+    let canister_info = CanisterInfo {
+        id: canister_id,
+        timestamp: canister_timestamp,
+    };
+    let agent = env.get_agent().context("Failed to get HTTP agent")?;
+    let playground_canister = match env.get_network_descriptor().r#type {
+        NetworkTypeDescriptor::Playground {
+            playground_canister,
+            ..
+        } => playground_canister,
+        _ => bail!("Trying to install wasm through playground on non-playground network."),
+    };
+    let install_arg = InstallArgs {
+        arg,
+        wasm_module,
+        mode,
+        canister_id: canister_info.id,
+    };
+    let encoded_arg = encode_args((canister_info, install_arg, false, is_asset_canister))?;
+    let result = agent
+        .update(&playground_canister, "installCode")
+        .with_arg(encoded_arg.as_slice())
+        .call_and_wait()
+        .await
+        .context("install failed")?;
+    let out = Decode!(&result, CanisterInfo)?;
+    let refreshed_timestamp = out.timestamp;
+    Ok(refreshed_timestamp.into())
+}
+
+fn create_nonce() -> (candid::Int, candid::Nat) {
+    let now = SystemTime::now()
+        .duration_since(SystemTime::UNIX_EPOCH)
+        .unwrap()
+        .as_millis();
+    let timestamp = candid::Int::from(now * 1_000_000);
+    let mut rng = rand::thread_rng();
+    let mut nonce = candid::Nat::from(rng.gen::<i32>());
+    let prefix = format!("{}{}", POW_DOMAIN, timestamp);
+    loop {
+        let to_hash = format!("{}{}", prefix, nonce).replace('_', "");
+        let hash = motoko_hash(&to_hash);
+        if (hash & 0xc0000000) == 0 {
+            return (timestamp, nonce);
+        }
+        nonce += 1;
```

Comment:
Is it okay if `nonce` ends up exceeding the largest value that `rng.gen::<i32>()` could have generated?

Reply:
Yes, the nonce may be anything in the whole `Nat` (that's a `BigUInt` behind the scenes) spectrum